dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataelement/hibernate/HibernateDataElementStore.java

```diff
@@ -145,9 +142,32 @@ public List<DataElement> getDataElementsByAggregationLevel(int aggregationLevel)
   public DataElement getDataElement(String uid, User user) {
     CriteriaBuilder builder = getCriteriaBuilder();
 
+    // Need to refetch here since the user might have been updated, and tests transactional
+    // semantics might not have committed yet.
+    CurrentUserGroupInfo currentUserGroupInfo = getCurrentUserGroupInfo(user.getUid());
+    if (user.getGroups().size() != currentUserGroupInfo.getUserGroupUIDs().size()) {
+      String msg =
+          String.format(
+              "User '%s' getGroups().size() has %d groups, but  getUserGroupUIDs() returns %d groups!",
+              user.getUsername(),
+              user.getGroups().size(),
+              currentUserGroupInfo.getUserGroupUIDs().size());
+
+      RuntimeException runtimeException = new RuntimeException(msg);
+      log.error(msg, runtimeException);
+      throw runtimeException;
+    }
+
+    List<Function<Root<DataElement>, Predicate>> sharingPredicates =
+        getSharingPredicates(
+            builder,
+            user.getUid(),
+            currentUserGroupInfo.getUserGroupUIDs(),
+            AclService.LIKE_READ_METADATA);
+
```

Comment:
would it make sense for the new code here (145 - 167) to be its own method so it doesn't muddy this method which gets data elements?

Reply:
My intention to remove the CurrentUserGroupInfo all together, but I need to do that in a separate PR. The check for size inconsistency was just a convenience during testing. I will remove that.